src/solver/expressions/include/antares/solver/expressions/nodes/NodeVisitor.h

```diff
@@ -0,0 +1,96 @@
+/*
+** Copyright 2007-2024, RTE (https://www.rte-france.com)
+** See AUTHORS.txt
+** SPDX-License-Identifier: MPL-2.0
+** This file is part of Antares-Simulator,
+** Adequacy and Performance assessment for interconnected energy networks.
+**
+** Antares_Simulator is free software: you can redistribute it and/or modify
+** it under the terms of the Mozilla Public Licence 2.0 as published by
+** the Mozilla Foundation, either version 2 of the License, or
+** (at your option) any later version.
+**
+** Antares_Simulator is distributed in the hope that it will be useful,
+** but WITHOUT ANY WARRANTY; without even the implied warranty of
+** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+** Mozilla Public Licence 2.0 for more details.
+**
+** You should have received a copy of the Mozilla Public Licence 2.0
+** along with Antares_Simulator. If not, see <https://opensource.org/license/mpl-2-0/>.
+*/
+#pragma once
+#include <array>
+#include <optional>
+#include <string>
+
+#include <antares/logs/logs.h>
+#include <antares/solver/expressions/nodes/NodesForwardDeclaration.h>
+
+namespace Antares::Solver::Expressions
+{
+
+template<class R>
+class NodeVisitor
+{
+private:
+    template<class T>
+    std::optional<R> tryType(const Node& node)
+    {
+        if (const T* x = dynamic_cast<const T*>(&node))
+        {
+            return visit(*x); // on appelle la version 'T' (virtuelle pure)
+        }
+        return {};
+    }
+
+public:
+    virtual ~NodeVisitor() = default;
+
+    R dispatch(const Node& node)
+    {
+        using Function = std::optional<R> (Antares::Solver::Expressions::NodeVisitor<R>::*)(
+          const Antares::Solver::Expressions::Node&);
+        static const std::array<Function, 16> tryFunctions{
```

Comment:
About `static const std::array<Function, 16> tryFunctions` : 
- This list of function objects does not have to be defined inside **dispatch** method, does it ? It bloats the **dispatch** function. where else could we define it ? Why not outside the class itself (but in the same header) ? Or as a private data member. 
- Does it have to be **static** ?
- Does it have to be a **std::array** ? A **std::vector** could be a tiny bit easier to define, as we don't have to give its size (unlike **std::array**). And in case of change, the size does not have to be updated (ok, not a big improvement...).
- This list contains **tryType** functions. 
  About these functions : technically, they do 2 things : 
  - they turn the argument's polymorphic type into the expected type (template parameter) 
  - they run **visit** function on the object of the expected type (if it is of the expected type)
  
  So, to me, name **tryType** is vague, and does describe what the function actually does.
  
About this last point, we could instead : 
Redefine **tryType** template function like this : 
```C++
template<class T>
T* tryType(const Node& node)
{
    return dynamic_cast<const T*>(&node);
}
```
and change a bit the **dispatch** function : 
```c++
R dispatch(const Node& node)
{
    ...
    for (auto f: tryTypes) // tryFunctions renamed into tryTypes
    {
        if (auto* actualNode = f(node))
        {
            return visit(*actualNode);
        }
    }

    logs.error() << "Antares::Solver::Expressions Visitor: unsupported Node!";
    return R();
}
```
It's more simple like this : 
- the **std::optional<R>** disappeared.
- template **tryType** is shorter and is named according to what it actually does

Reply:
> About `static const std::array<Function, 16> tryFunctions` :
> 
>     * This list of function objects does not have to be defined inside **dispatch** method, does it ? It bloats the **dispatch** function. Where else could we define it ? Why not outside the class itself (but in the same header) ? Or as a private data member.
> 
Could be outside
>     * Does it have to be **static** ?
>
Yes, we don't want to rebuild it at every function call (1000s of function calls) for obvious reasons 
>     * Does it have to be a **std::array** ? A **std::vector** could be a tiny bit easier to define, as we don't have to give its size (unlike **std::array**). And in case of change, the size does not have to be updated (ok, not a big improvement...).
> 
Yes it can be a `std::vector`
>     * This list contains **tryType** functions.
>       About these functions : technically, they do 2 things :
>       
>       * they turn the argument's polymorphic type into the expected type (template parameter)
>       * they run **visit** function on the object of the expected type (if it is of the expected type)
>       
>       So, to me, name **tryType** is vague, and does describe what the function actually does.
> 
> 
> About this last point, we could instead : Redefine **tryType** template function like this :
> 
> ```c++
> template<class T>
> T* tryType(const Node& node)
> {
>     return dynamic_cast<const T*>(&node);
> }
> ```
> 
> and change a bit the **dispatch** function :
> 
> ```c++
> R dispatch(const Node& node)
> {
>     ...
>     for (auto f: tryTypes) // tryFunctions renamed into tryTypes
>     {
>         if (auto* actualNode = f(node))
>         {
>             return visit(*actualNode);
>         }
>     }
> 
>     logs.error() << "Antares::Solver::Expressions Visitor: unsupported Node!";
>     return R();
> }
> ```
> 
> It's more simple like this :
> 
>     * the **std::optional** disappeared.
> 
>     * template **tryType** is shorter and is named according to what it actually does

I agree that your code is simpler. But it doesn't compile because you can't (easily at least) iterate over types in C++.